cert/aziot-certd/src/lib.rs

```diff
@@ -195,654 +258,349 @@ impl UpdateConfig for Api {
+                            )?;
 
-                    let x509 = create_cert(api, id, csr, Some((&issuer_cert, &issuer_private_key)))
-                        .await?;
-                    Ok(x509)
-                }
+                        if let Ok(ref handle) = api.key_client.load_key_pair(&auth_x509.identity.pk) {
+                            api.key_client.delete_key_pair(handle)?;
+                        }
 
-                CertIssuanceMethod::SelfSigned => {
-                    // Since the client did not give us their private key handle, we assume that the key is named the same as the cert.
-                    //
-                    // TODO: Is there a way to not have to assume this?
-                    let key_pair_handle = api
-                        .key_client
-                        .load_key_pair(id)
-                        .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-
-                    // Recurse with explicit issuer.
-                    let x509 = create_cert(api, id, csr, Some((id, &key_pair_handle))).await?;
-                    Ok(x509)
+                        let handle = api.key_client.create_key_pair_if_not_exists(
+                                &auth_x509.identity.pk,
+                                Some("ec-p256:rsa-4096:*")
+                            )
+                            .and_then(|handle| Ok(CString::new(handle.0)?))?;
+
+                        let id_pubkey = api.key_engine.load_public_key(&handle)?;
+                        let id_privkey = api.key_engine.load_private_key(&handle)?;
+
+                        let subject_name = build_name(
+                                cert_options.subject.as_ref()
+                                    .unwrap_or(&CertSubject::CommonName("est-id".to_owned()))
+                            )?;
+
+                        let mut builder = X509Req::builder()?;
+                        builder.set_version(0)?;
+                        builder.set_subject_name(&subject_name)?;
+
+                        let mut exts = Stack::new()?;
+                        exts.push(
+                                extension::ExtendedKeyUsage::new()
+                                    .client_auth()
+                                    .build()?
+                            )?;
+
+                        builder.add_extensions(&exts)?;
+                        builder.set_pubkey(&id_pubkey)?;
+                        builder.sign(&id_privkey, MessageDigest::sha256())?;
+
+                        let id_init = est::create_cert(
+                                base64::encode(builder.build().to_der()?).into_bytes(),
+                                url,
+                                auth.basic.as_ref(),
+                                Some((&bcert, &bpk)),
+                                &trusted_certs,
+                                api.proxy_uri.clone()
+                            )
+                            .await?;
```

Comment:
Restore the error message from the original code. It contained both the error from getting the EST ID cert and the error from issuing a new one.

Reply:
Fixed in 56ac22b39cba8804ead30ca1d3031387e08366b9.

https://github.com/Azure/iot-identity-service/pull/296/commits/56ac22b39cba8804ead30ca1d3031387e08366b9#diff-ce1948fdb394a64e8a42f87d92e2057992c63a20520bad0c4c7359b64d0854c2R570-R578